allennlp/semparse/executors/wikitables_variable_free_executor.py

```diff
@@ -0,0 +1,647 @@
+        returns a list of tuples, each containing as the first element a date taken from
+        that column, and the corresponding row as the second element. The output can be used to
+        compare rows based on the dates.
+        """
+        row_list, column_name = self._get_row_list_and_column_name(expression_list)
+        if not row_list:
+            return []
+        cell_row_pairs = [(self._make_date(row[column_name].replace('fb:cell.', '')), row)
+                          for row in row_list]
+        return cell_row_pairs
+
+    @staticmethod
+    def _make_date(cell_string: str) -> Date:
+        string_parts = cell_string.split("_")
+        year = -1
+        month = -1
+        day = -1
+        for part in string_parts:
+            if part.isdigit():
+                if len(part) == 4:
+                    year = int(part)
+                else:
+                    day = int(part)
+            elif part in MONTH_NUMBERS:
+                month = MONTH_NUMBERS[part]
+        return Date(year, month, day)
+
+    def _values_look_like_dates(self, expression_list: NestedList) -> bool:
+        # TODO (pradeep): This will be unnecessary when we have column types identified.
+        row_list, column_name = self._get_row_list_and_column_name(expression_list)
+        if not row_list:
+            return False
+        # We try to figure out if the values being compared are simple numbers or dates. We use
+        # simple rules here: that the string contains less than 4 parts, and one of the parts is a
+        # month name. Note that this will not consider strings with just years as dates. That's fine
+        # because we can compare them as numbers.
+        values_are_dates = False
+        # Apply rules on the first cell only, and assume the decision applies to the remaining cells
+        # as well.
+        cell_value_parts = row_list[0][column_name].replace('fb:cell.', '').split('_')
+        # Check if the number of parts in the string are 3 or fewer. If not, it's probably neither a
+        # date nor a number.
+        if len(cell_value_parts) <= 3:
+            for part in cell_value_parts:
+                if part in MONTH_NUMBERS:
+                    values_are_dates = True
+        return values_are_dates
+
+    def _argmax(self, expression_list: NestedList) -> List[Dict[str, str]]:
+        """
+        Takes a list of rows and a column (decoded from the `expression_list`) and returns a list
+        containing a single row (dict from columns to cells) that has the maximum numerical value in
+        the given column. We return a list instead of a single dict to be consistent with the return
+        type of `_select` and `_all_rows`.
+        """
+        if self._values_look_like_dates(expression_list):
+            date_row_pairs = self._get_date_row_pairs_to_filter(expression_list)
+            if not date_row_pairs:
```

Comment:
Looks like you can pull out the last four lines of this, and just have

```python
if self._values_look_like_dates(row_expression):
    row_pairs = ...
else:
    row_pairs = ...
if not row_pairs:
    return []
return [sorted(...)]
```

Reply:
mypy complains if values of different types are assigned to the same variable in if and else blocks.

Reply:
Can you do something with defining `row_pairs: SomeType` above, with a more generic type than it would get below?  If not, ignoring mypy seems reasonable, as you really shouldn't be duplicating the logic here.

Reply:
I tried a couple of different options, including defining a generic type before the `if` block, but mypy was unforgiving. I'll ignore it for now. But this seems like a common coding pattern. There should be a better solution.